rhine/bench/WordCount.hs

```diff
@@ -0,0 +1,126 @@
+-- | Count the number of words in the complete works of Shakespeare.
+
+{-# LANGUAGE Arrows #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+
+module WordCount where
+
+-- base
+import Control.Exception
+import Data.IORef (IORef, modifyIORef', newIORef, readIORef, writeIORef)
+import Data.Monoid (Sum (..))
+import GHC.IO.Handle hiding (hGetContents)
+import System.IO (IOMode (ReadMode), openFile, stdin, withFile)
+import System.IO.Error (isEOFError)
+import Prelude hiding (getContents, getLine, words)
+
+-- text
+import Data.Text (words)
+import Data.Text.IO (getLine)
+import Data.Text.Lazy qualified as Lazy
+import Data.Text.Lazy.IO (hGetContents)
+
+-- criterion
+import Criterion.Main
+
+-- dunai
+import Data.MonadicStreamFunction qualified as Dunai
+
+-- rhine
+import FRP.Rhine
+import Paths_rhine
+
+benchmarks :: Benchmark
+benchmarks =
+  bgroup
+    "WordCount"
+    [ bench "rhine" $ nfIO rhineWordCount
+    , bench "dunai" $ nfIO dunaiWordCount
+    , bgroup
+        "Text"
+        [ bench "IORef" $ nfIO textWordCount
+        , bench "no IORef" $ nfIO textWordCountNoIORef
+        , bench "Lazy" $ nfIO textLazy
+        ]
+    ]
+
+testFile :: IO FilePath
+testFile = getDataFileName "bench/pg100.txt"
+
+withInput :: IO b -> IO b
+withInput action = do
+  inputFileName <- testFile
+  withFile inputFileName ReadMode $ \stdinFile -> do
+    hDuplicateTo stdinFile stdin
+    action
+
+-- FIXME the StdinClock should really throw something in Except instead.
```

Comment:
Maybe I should finish #287 first?

Reply:
Yes, it should be finished first, and then I can look into adding both variants here. If they turn out to have the same performance, I can keep the more idiomatic one.